README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# lubridateExtras <img src="man/figures/logo.svg" align="right" height="120" width="139" />

Convenience functions for the lubridate package

<!-- badges: start -->

[![R build
status](https://github.com/ellisvalentiner/lubridateExtras/workflows/R-CMD-check/badge.svg)](https://github.com/ellisvalentiner/lubridateExtras/actions)
[![Coverage](%5B!%5Bcodecov%5D(https://codecov.io/gh/ellisvalentiner/lubridateExtras/branch/main/graph/badge.svg?token=G4DGE1C5dG)%5D(https://codecov.io/gh/ellisvalentiner/lubridateExtras))](https://app.codecov.io/gh/ellisvalentiner/lubridateExtras/)
<!-- badges: end -->

## Overview

Lubridate makes it easier to work with date-time data in R and provides
new capabilities. LubridateExtras builds on top of lubridate to provide
a number of convenience functions, primarily focused on abstracting
patterns in ways that improve code readability and reduce copying and
pasting code.

## Installation

``` r
# lubridateExtras is not currently on CRAN
# Please install the development version from GitHub:
# install.packages("devtools")
devtools::install_github("ellisvalentiner/lubridateExtras")
```

If you encounter a clear bug, please file a minimal reproducible example
on [github](https://github.com/ellisvalentiner/lubridateExtras/issues).

## Usage

``` r
library(lubridateExtras)

yesterday()
#> [1] "2022-01-23"

tomorrow()
#> [1] "2022-01-25"

days_ago(7)  # equivalent to lubridate::today() - lubridate::days(7)
#> [1] "2022-01-17"

days_hence(7)  # equivalent to lubridate::today() + lubridate::days(7)
#> [1] "2022-01-31"

hms("2017-10-22 15:33:00")  # extracts the time-of-day component
#> [1] "15H 33M 0S"
```


NEWS.md

# lubridateExtras 0.1.4

## 2022-01-24

* Minor fixes and improvements

## 2017-09-24

* Initial release notes


tests/testthat.R
library(testthat)
library(lubridateExtras)

test_check("lubridateExtras")


tests/testthat/test_lint.R
if (requireNamespace("lintr", quietly = TRUE)) {
  library(lintr)
  context("Lints")
  test_that("Package Style", {
    # A duplicate copy of the find_package function from lintr
    find_package <- function(path) {
      path <- normalizePath(path, mustWork = FALSE)

      while (!file.exists(file.path(path, "DESCRIPTION"))) {
        path <- dirname(path)
        if (identical(path, dirname(path))) {
          return(NULL)
        }
      }

      path
    }

    if (!is.null(find_package("."))) {
      expect_lint_free()
    }
  })
}


tests/testthat/test_round.R
library(lubridate)
library(lubridateExtras)
context("Round Functions")

test_that("round_hms", {
  x <- lubridate::hms("12:34:56")
  expect_equal(round_hms(x, "5 seconds"), lubridate::hms("12:34:55"))
  expect_equal(round_hms(x, "5 minutes"), lubridate::hms("12:35:00"))
})

test_that("floor_hms", {
  x <- lubridate::hms("12:34:56")
  expect_equal(floor_hms(x, "5 seconds"), lubridate::hms("12:34:55"))
  expect_equal(floor_hms(x, "5 minutes"), lubridate::hms("12:30:00"))
})

test_that("ceiling_hms", {
  x <- lubridate::hms("12:34:56")
  expect_equal(ceiling_hms(x, "5 seconds"), lubridate::hms("12:35:00"))
  expect_equal(ceiling_hms(x, "5 minutes"), lubridate::hms("12:35:00"))
})

test_that("first_*", {
  x <- lubridate::as_datetime("2017-10-22 15:01:00")
  expect_equal(first_monday(x), lubridate::as_date("2017-10-02"))
  expect_equal(first_monday(x, "year"), lubridate::as_date("2017-01-02"))
  expect_equal(first_tuesday(x), lubridate::as_date("2017-10-03"))
  expect_equal(first_tuesday(x, "year"), lubridate::as_date("2017-01-03"))
})


R/zzz.R
# nocov start

.onLoad <- function(libname, pkgname) {
  ## source: https://github.com/tidyverse/lubridate/blob/master/R/zzz.R#L3-L25
  if (Sys.getenv("TZDIR") == "") {
    ## adapted from OlsonNames function
    tzdir <-
      if (.Platform$OS.type == "windows") {
        file.path(R.home("share"), "zoneinfo")
      } else if (!file.exists("/usr/share/zoneinfo")) {
        tzdirs <- c(
          file.path(R.home("share"), "zoneinfo"),
          "/usr/share/lib/zoneinfo", "/usr/lib/zoneinfo",
          "/usr/local/etc/zoneinfo", "/etc/zoneinfo",
          "/usr/etc/zoneinfo"
        )
        tzdirs <- tzdirs[file.exists(tzdirs)]
        if (length(tzdirs)) {
          tzdirs[[1]]
        } else {
          NULL
        }
      }
    if (!is.null(tzdir)) {
      Sys.setenv(TZDIR = tzdir)
    }
  }
}

# nocov end


R/round.R
#' Round, floor and ceiling methods for hms objects
#'
#' @description
#' Rounding to the nearest unit or multiple of a unit are supported. All
#' meaningful specifications in English language are supported - secs, min,
#' mins, 2 minutes, etc.
#'
#' Rounding to fractional seconds is supported. Please note that rounding to
#' fractions smaller than 1s can lead to large precision errors due to the
#' floating point representation of the POSIXct objects. See examples.
#'
#' `round_date()` takes a hms object and rounds it to the nearest value
#' of the specified time unit. For rounding hms which is exactly halfway
#' between two consecutive units, the convention is to round up. Note that this
#' is in line with the behavior of R's [base::round.POSIXt()] function
#' but does not follow the convention of the base [base::round()] function
#' which "rounds to the even digit" per IEC 60559.
#'
#' @rdname round_hms
#' @param x a vector of date-time objects
#' @param unit a character string specifying a time unit or a multiple of a unit
#'   to be rounded to. Valid base units are `second`, `minute`, and `hour`.
#'   Arbitrary unique English abbreviations as in the [period()]
#'   constructor are allowed. Rounding to multiple of units is supported.
#' @keywords manip chron
#' @seealso [base::round(), lubridate::round_date(), lubridate::floor_date(),
#'   lubridate::ceiling_date()]
#' @examples
#'
#' ## print fractional seconds
#' options(digits.secs = 6)
#'
#' x <- lubridate::hms("12:01:59.23")
#' round_hms(x, ".5s")
#' round_hms(x, "sec")
#' round_hms(x, "second")
#' round_hms(x, "minute")
#' round_hms(x, "5 mins")
#' round_hms(x, "hour")
#' round_hms(x, "2 hours")
#'
#' x <- lubridate::hms("12:01:59.23")
#' floor_hms(x, "12s")
#' floor_hms(x, "second")
#' floor_hms(x, "minute")
#' floor_hms(x, "hour")
#'
#' x <- lubridate::hms("12:01:59.23")
#' ceiling_hms(x, "15 sec")
#' ceiling_hms(x, "second")
#' ceiling_hms(x, "minute")
#' ceiling_hms(x, "5 mins")
#' ceiling_hms(x, "hour")
#' @export
round_hms <- function(x, unit = "second") {
  if (!length(x)) {
    return(x)
  }
  x |>
    lubridate::hms() |>
    lubridate::as_datetime() |>
    lubridate::round_date(unit = unit) |>
    lubridateExtras::hms()
}

#' @description
#' `floor_hms()` takes a hms object and rounds it down to the nearest
#' boundary of the specified time unit.
#' @rdname round_hms
#' @export
floor_hms <- function(x, unit = "seconds") {
  if (!length(x)) {
    return(x)
  }
  x |>
    lubridate::hms() |>
    lubridate::as_datetime() |>
    lubridate::floor_date(unit = unit) |>
    lubridateExtras::hms()
}
#' @description
#' `ceiling_hms()` takes a hms object and rounds it up to the nearest
#' boundary of the specified time unit.
#' @rdname round_hms
#' @export
#' @examples
#' x <- lubridate::hms("12:01:59.23")
#' ceiling_hms(x, "10 seconds")
ceiling_hms <- function(x, unit = "seconds") {
  if (!length(x)) {
    return(x)
  }
  x |>
    lubridate::hms() |>
    lubridate::as_datetime() |>
    lubridate::ceiling_date(unit = unit) |>
    lubridateExtras::hms()
}


R/instants.R
#' The previous day
#'
#' @export yesterday
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the previous date as a Date object
#'
#' @seealso \code{\link{tomorrow}}
#'
#' @examples
#' yesterday()
#' yesterday("UTC")
yesterday <- function(tzone = "") {
  as_date(now(tzone = tzone) - days(1))
}

#' The next day
#'
#' @export tomorrow
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the previous date as a Date object
#'
#' @seealso \code{\link{yesterday}}
#'
#' @examples
#' tomorrow()
#' tomorrow("UTC")
tomorrow <- function(tzone = "") {
  as_date(now(tzone = tzone) + days(1))
}

#' The time x seconds before now
#'
#' @param seconds integer number of seconds
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The datetime, x seconds ago
#' @export seconds_ago
#'
#' @examples
#' seconds_ago(120)
#' seconds_ago(3600)
seconds_ago <- function(seconds = 0, tzone = "") {
  as_datetime(now() - seconds(x = seconds), tz = tzone)
}

#' The time x minutes before now
#'
#' @param minutes integer number of minutes
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The time, x minutes ago
#' @export minutes_ago
#'
#' @examples
#' minutes_ago(1)
#' minutes_ago(60)
minutes_ago <- function(minutes = 0, tzone = "") {
  as_datetime(now() - minutes(x = minutes), tz = tzone)
}

#' the time x hours ago
#'
#' @param hours integer number of hours
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The time, x hours ago
#' @export hours_ago
#'
#' @examples
#' hours_ago(1)
#' hours_ago(24)
hours_ago <- function(hours = 0, tzone = "") {
  as_datetime(now() - hours(x = hours), tz = tzone)
}

#' The date x days ago
#'
#' @export days_ago
#' @param days an integer specifying the number of days to subtract from the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the date, x days ago
#'
#' @examples
#' days_ago(3)
#' days_ago(1) == yesterday()
days_ago <- function(days = 0, tzone = "") {
  as_date(today(tzone = tzone) - days(x = days))
}

#' The date x weeks ago
#'
#' @param weeks an integer specifying the number of weeks to subtract from the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return the date, x weeks ago
#' @export weeks_ago
#'
#' @examples
#' weeks_ago(1)
#' weeks_ago(52)
weeks_ago <- function(weeks = 0, tzone = "") {
  as_date(today(tzone = tzone) - weeks(x = weeks))
}


#' The date x months ago
#'
#' @param months an integer specifying th enumber of weeks to subtract from the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return the date, x months ago
#' @export months_ago
#'
#' @examples
#' months_ago(1)
#' months_ago(12)
months_ago <- function(months = 0, tzone = "") {
  as_date(today(tzone = tzone) - months(x = months))
}

#' The date x years ago
#'
#' @param years an integer specifying the number of days to subtract from the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The date, x years ago
#' @export years_ago
#'
#' @examples
#' years_ago(1)
#' years_ago(2000)
years_ago <- function(years = 0, tzone = "") {
  as_date(today(tzone = tzone) - years(x = years))
}

#' The date x decades ago
#'
#' @param decades an integer specifying the number of decades to subtract
#'   from the current date
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The date, x decades ago
#' @export decades_ago
#'
#' @examples
#' decades_ago(1)
#' decades_ago(2000)
decades_ago <- function(decades = 0, tzone = "") {
  as_date(today(tzone = tzone) - years(x = decades * 10))
}

#' The time x seconds before now
#'
#' @param seconds integer number of seconds
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The datetime, x seconds from now
#' @export seconds_hence
#'
#' @examples
#' seconds_hence(60)
#' seconds_hence(3600)
seconds_hence <- function(seconds = 0, tzone = "") {
  as_datetime(now() + seconds(x = seconds), tz = tzone)
}

#' The time x minutes before now
#'
#' @param minutes integer number of minutes
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The date time, x minutes from now
#' @export minutes_hence
#'
#' @examples
#' minutes_hence(60)
#' minutes_hence(2440)
minutes_hence <- function(minutes = 0, tzone = "") {
  as_datetime(now() + minutes(x = minutes), tz = tzone)
}

#' the time x hours hence
#'
#' @param hours integer number of hours
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The datetime, x hours from now
#' @export hours_hence
#'
#' @examples
#' hours_hence(1)
#' hours_hence(24)
hours_hence <- function(hours = 0, tzone = "") {
  as_datetime(now() + hours(x = hours), tz = tzone)
}
#' The date x days hence
#'
#' @export days_hence
#' @param days an integer specifying the number of days to add from the current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the datetime, x days from now
#'
#' @examples
#' days_hence(3)
#' days_hence(1) == tomorrow()
days_hence <- function(days = 0, tzone = "") {
  as_date(today(tzone = tzone) + days(x = days))
}

#' The current month
#'
#' @export this_month
#' @param tzone a character vector specifying which time zone you would like to
#'   find the current month of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the current month as a Date object
this_month <- function(tzone = "") {
  floor_date(x = today(tzone = tzone), unit = "month")
}

#' The date x weeks hence
#'
#' @param weeks an integer specifying the number of weeks to add to the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return the date, x weeks from now
#' @export weeks_hence
#'
#' @examples
#' weeks_hence(1)
#' weeks_hence(52)
weeks_hence <- function(weeks = 0, tzone = "") {
  as_date(today(tzone = tzone) + weeks(x = weeks))
}


#' The date x months hence
#'
#' @param months an integer specifying the number of weeks to add to the current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return the date, x months from now
#' @export months_hence
#'
#' @examples
#' months_hence(1)
#' months_hence(12)
months_hence <- function(months = 0, tzone = "") {
  as_date(today(tzone = tzone) + months(x = months))
}

#' The date x years hence
#'
#' @param years an integer specifying the number of days to add to the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The date, x years from now
#' @export years_hence
#'
#' @examples
#' years_hence(1)
#' years_hence(2000)
years_hence <- function(years = 0, tzone = "") {
  as_date(today(tzone = tzone) + years(x = years))
}

#' The date x decades hence
#'
#' @param decades an integer specifying the number of decades to add to the
#'   current
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous date of. tzone defaults to the system time zone set on
#'   your computer.
#'
#' @return The date, x decades from now
#' @export decades_hence
#'
#' @examples
#' decades_hence(1)
#' decades_hence(2000)
decades_hence <- function(decades = 0, tzone = "") {
  as_date(today(tzone = tzone) + years(x = decades * 10))
}

#' The previous month
#'
#' @export last_month
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous month of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the previous month as a Date object
last_month <- function(tzone = "") {
  floor_date(x = today(tzone = tzone), unit = "month") - months(1)
}

#' The next month
#'
#' @export next_month
#' @param tzone a character vector specifying which time zone you would like to
#'   find the next month of. tzone defaults to the system time zone set on your
#'   computer.
#' @return the next month as a Date object
next_month <- function(tzone = "") {
  floor_date(x = today(tzone = tzone), unit = "month") + months(1)
}

#' The current week
#'
#' @export this_week
#' @param tzone a character vector specifying which time zone you would like to
#'   find the current week of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the current week as a Date object
this_week <- function(tzone = "") {
  floor_date(x = today(tzone = tzone), unit = "week")
}

#' The previous week
#'
#' @export last_week
#' @param tzone a character vector specifying which time zone you would like to
#'   find the previous week of. tzone defaults to the system time zone set on
#'   your computer.
#' @return the previous week as a Date object
last_week <- function(tzone = "") {
  floor_date(x = today(tzone = tzone), unit = "week") - weeks(1)
}

#' The next week
#'
#' @export next_week
#' @param tzone a character vector specifying which time zone you would like to
#'   find the next week of. tzone defaults to the system time zone set on your
#'   computer.
#' @return the next month as a Date object
next_week <- function(tzone = "") {
  floor_date(x = today(tzone = tzone), unit = "week") + weeks(1)
}

#' Is x a weekend?
#'
#' @export is_weekend
#' @param x a POSIXct, POSIXlt, Date, chron, yearmon, yearqtr, zoo, zooreg,
#'   timeDate, xts, its, ti, jul, timeSeries, or fts object.
#' @return boolean indicating whether x is a weekend
#'
#' @examples
#' is_weekend("2017-08-29") # FALSE
#' is_weekend("2017-09-02") # TRUE
is_weekend <- function(x) {
  wday(x = as_date(x), label = FALSE, abbr = FALSE) %in% c(1, 7)
}

#' Is x a weekday?
#'
#' @export is_weekday
#' @param x a POSIXct, POSIXlt, Date, chron, yearmon, yearqtr, zoo, zooreg,
#'   timeDate, xts, its, ti, jul, timeSeries, or fts object.
#' @return boolean indicating whether x is a weekday
#'
#' @examples
#' is_weekday("2017-08-29") # FALSE
#' is_weekday("2017-09-02") # TRUE
is_weekday <- function(x) {
  wday(x = as_date(x), label = FALSE, abbr = FALSE) %in% 2:6
}

#' Extract time-of-day values
#'
#' @export hms
#' @param x a POSIXct, POSIXlt, Date, chron, yearmon, yearqtr, zoo, zooreg,
#'   timeDate, xts, its, ti, jul, timeSeries, or fts object.
#' @return hms object
#' @examples
#' hms("2017-10-22 15:01:00")
hms <- function(x) {
  x <- parse_date_time(x = x, orders = c("ymdTz", "ymdT", "HMS"))
  x <- strftime(x, format = "%H:%M:%S", tz = "GMT")
  lubridate::hms(x)
}

#' The beginning (floor) of the decade
#'
#' @export floor_decade
#' @param x a POSIXct, POSIXlt, Date, chron, yearmon, yearqtr, zoo, zooreg,
#'   timeDate, xts, its, ti, jul, timeSeries, or fts object.
#' @return the beginning of the decade as a Date object
#'
#' @examples
#' floor_decade("2017-10-22 15:01:00")
floor_decade <- function(x) {
  x <- parse_date_time(x = x, orders = c("ymd", "ymdTz", "ymdT"))
  x <- floor_date(x, unit = "year")
  beginning <- year(x) - year(x) %% 10
  year(x) <- beginning
  return(x)
}

#' The end (ceiling) of a decade
#'
#' @export ceiling_decade
#' @param x a POSIXct, POSIXlt, Date, chron, yearmon, yearqtr, zoo, zooreg,
#'   timeDate, xts, its, ti, jul, timeSeries, or fts object.
#' @return the end of the decade as a Date object
#'
#' @examples
#' ceiling_decade("2017-10-22 15:01:00")
ceiling_decade <- function(x) {
  x <- parse_date_time(x = x, orders = c("ymd", "ymdTz", "ymdT"))
  decade_end <- year(x) - (year(x) %% 10) + 10
  year(x) <- decade_end
  x <- floor_date(x, unit = "year") - days(1)
  return(x)
}
